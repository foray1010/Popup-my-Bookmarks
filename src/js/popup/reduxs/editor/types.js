// @flow strict

export type State = {|
  isAllowEditUrl: boolean,
  isCreating: boolean,
  positionLeft: number,
  positionTop: number,
  targetId: string,
  title: string,
  url: string
|}
