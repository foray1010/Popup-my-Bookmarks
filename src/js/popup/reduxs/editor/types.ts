export interface EditorState {
  isAllowEditUrl: boolean
  isCreating: boolean
  positionLeft: number
  positionTop: number
  targetId: string
  title: string
  url: string
}
