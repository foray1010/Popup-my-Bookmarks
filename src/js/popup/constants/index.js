// @flow strict

export * from '../../common/constants'

export const GOLDEN_GAP = 2
export const MAX_HEIGHT = 598
export const MIN_BOOKMARK_ICON_SIZE = 16
export const NO_BOOKMARK_ID_PREFIX = 'NO_BOOKMARK_ID_'
export const SEARCH_RESULT_ID = 'SEARCH_RESULT_ID'
export const SEPARATE_THIS_URL = 'http://separatethis.com/'

export const MENU_ADD_FOLDER = 'addFolder'
export const MENU_ADD_PAGE = 'addPage'
export const MENU_ADD_SEPARATOR = 'addSeparator'
export const MENU_COPY = 'copy'
export const MENU_CUT = 'cut'
export const MENU_DEL = 'del'
export const MENU_EDIT = 'edit'
export const MENU_OPEN_ALL = 'openAll'
export const MENU_OPEN_ALL_IN_I = 'openAllInI'
export const MENU_OPEN_ALL_IN_N = 'openAllInN'
export const MENU_OPEN_IN_B = 'openInB'
export const MENU_OPEN_IN_I = 'openInI'
export const MENU_OPEN_IN_N = 'openInN'
export const MENU_PASTE = 'paste'
export const MENU_RENAME = 'rename'
export const MENU_SORT_BY_NAME = 'sortByName'

export const OPEN_IN_TYPES = {
  BACKGROUND_TAB: 'BACKGROUND_TAB',
  CURRENT_TAB: 'CURRENT_TAB',
  INCOGNITO_WINDOW: 'INCOGNITO_WINDOW',
  NEW_TAB: 'NEW_TAB',
  NEW_WINDOW: 'NEW_WINDOW'
}

export const TYPE_BOOKMARK = 'bookmark'
export const TYPE_DRAG_INDICATOR = 'drag-indicator'
export const TYPE_FOLDER = 'folder'
export const TYPE_NO_BOOKMARK = 'no-bookmark'
export const TYPE_SEPARATOR = 'separator'
