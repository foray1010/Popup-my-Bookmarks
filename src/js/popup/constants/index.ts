export * from '../../common/constants'

export const GOLDEN_GAP = 2
export const MAX_HEIGHT = 598
export const MIN_BOOKMARK_ICON_SIZE = 16
export const NO_BOOKMARK_ID_PREFIX = 'NO_BOOKMARK_ID_'
export const SEARCH_RESULT_ID = 'SEARCH_RESULT_ID'
export const SEPARATE_THIS_URL = 'http://separatethis.com/'

export const MENU_ADD_FOLDER = 'addFolder'
export const MENU_ADD_PAGE = 'addPage'
export const MENU_ADD_SEPARATOR = 'addSeparator'
export const MENU_COPY = 'copy'
export const MENU_CUT = 'cut'
export const MENU_DEL = 'del'
export const MENU_EDIT = 'edit'
export const MENU_OPEN_ALL = 'openAll'
export const MENU_OPEN_ALL_IN_I = 'openAllInI'
export const MENU_OPEN_ALL_IN_N = 'openAllInN'
export const MENU_OPEN_IN_B = 'openInB'
export const MENU_OPEN_IN_I = 'openInI'
export const MENU_OPEN_IN_N = 'openInN'
export const MENU_PASTE = 'paste'
export const MENU_RENAME = 'rename'
export const MENU_SORT_BY_NAME = 'sortByName'

export enum OPEN_IN_TYPES {
  BACKGROUND_TAB,
  CURRENT_TAB,
  INCOGNITO_WINDOW,
  NEW_TAB,
  NEW_WINDOW
}

export enum BOOKMARK_TYPES {
  BOOKMARK,
  DRAG_INDICATOR,
  FOLDER,
  NO_BOOKMARK,
  SEPARATOR
}
