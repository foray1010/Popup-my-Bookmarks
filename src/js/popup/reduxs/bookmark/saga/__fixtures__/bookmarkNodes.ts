import Chance from 'chance'
import * as R from 'ramda'

const chance = Chance('bookmarkNodes')

const generateBookmarkNode = () => ({
  id: String(chance.integer({min: 1000, max: 9999})),
  ...(chance.bool() ? {parentId: String(chance.integer({min: 1000, max: 9999}))} : null),
  ...(chance.bool() ? {index: chance.integer({min: 1000, max: 9999})} : null),
  ...(chance.bool() ? {url: chance.url()} : null),
  title: chance.word(),
  ...(chance.bool() ? {dateAdded: chance.integer({min: 1500000000000, max: 2000000000000})} : null),
  ...(chance.bool() ?
    {
      dateGroupModified: chance.integer({min: 1500000000000, max: 2000000000000})
    } :
    null),
  ...(chance.bool() ? {unmodifiable: chance.word()} : null)
})

export default R.times(generateBookmarkNode, 100)
