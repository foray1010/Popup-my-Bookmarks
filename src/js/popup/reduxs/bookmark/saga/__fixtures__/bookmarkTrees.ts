import Chance from 'chance'
import * as R from 'ramda'

import * as CST from '../../../../constants'

const chance = Chance('bookmarkTrees')

const generateType = () =>
  chance.pickone([
    CST.BOOKMARK_TYPES.BOOKMARK,
    CST.BOOKMARK_TYPES.FOLDER,
    CST.BOOKMARK_TYPES.SEPARATOR
  ])

const generateBookmarkInfo = (type?: CST.BOOKMARK_TYPES) => ({
  id: String(chance.integer({min: 1000, max: 9999})),
  parentId: String(chance.integer({min: 1000, max: 9999})),
  title: chance.word(),
  url: chance.url(),
  iconUrl: chance.url(),
  storageIndex: chance.integer({min: -1, max: 9999}),
  type: type || generateType(),
  isRoot: chance.bool(),
  isSimulated: false,
  isUnmodifiable: chance.bool()
})

export const generateBookmarkTree = () => ({
  children: R.times(() => generateBookmarkInfo(), 20),
  parent: generateBookmarkInfo(CST.BOOKMARK_TYPES.FOLDER)
})

export default R.times(generateBookmarkTree, 10)
