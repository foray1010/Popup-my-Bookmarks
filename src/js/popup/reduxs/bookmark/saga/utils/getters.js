// @flow strict

import * as R from 'ramda'
import type {Saga} from 'redux-saga'
import {all, call} from 'redux-saga/effects'

import type {Options} from '../../../../../common/types/options'
import {
  getBookmarkChildNodes,
  getBookmarkNodes,
  getI18n,
  searchBookmarkNodes
} from '../../../../../common/utils'
import * as CST from '../../../../constants'
import type {BookmarkInfo, BookmarkTree} from '../../../../types'
import {simulateBookmark, toBookmarkInfo} from '../../utils/converters'

export function* getBookmarkInfo(id: string): Saga<BookmarkInfo> {
  if (id.startsWith(CST.NO_BOOKMARK_ID_PREFIX)) {
    return simulateBookmark({
      id,
      parentId: id.replace(CST.NO_BOOKMARK_ID_PREFIX, ''),
      title: yield call(getI18n, 'noBkmark'),
      type: CST.TYPE_NO_BOOKMARK
    })
  }

  const bookmarkNodes = yield call(getBookmarkNodes, id)
  return toBookmarkInfo(bookmarkNodes[0])
}

export function* getBookmarkChildren(id: string): Saga<Array<BookmarkInfo>> {
  const bookmarkChildNodes = yield call(getBookmarkChildNodes, id)
  return R.map(toBookmarkInfo, bookmarkChildNodes)
}

export function* getBookmarkTree(id: string): Saga<BookmarkTree> {
  const [bookmarkInfo, bookmarkChildren] = yield all([
    call(getBookmarkInfo, id),
    call(getBookmarkChildren, id)
  ])
  return {
    children: bookmarkChildren.length ?
      bookmarkChildren :
      [yield call(getBookmarkInfo, CST.NO_BOOKMARK_ID_PREFIX + id)],
    parent: bookmarkInfo
  }
}

export function* getBookmarkTrees(
  restTreeIds: Array<string>,
  options: Options
): Saga<Array<BookmarkTree>> {
  const [firstTree, ...restTrees]: Array<BookmarkTree> = yield all([
    call(getFirstBookmarkTree, options),
    ...restTreeIds.map((id) => call(tryGetBookmarkTree, id))
  ])

  let acc = [firstTree]
  for (const tree of restTrees) {
    // if `tree` is deleted or not belong to this parent anymore,
    // ignore all its children
    const isBreak =
      tree === null ||
      // in case it is root folder that open from root
      (!tree.parent.isRoot && acc[acc.length - 1].parent.id !== tree.parent.parentId)
    if (isBreak) break

    acc = [...acc, tree]
  }
  return acc
}

export function* getFirstBookmarkTree(options: Options): Saga<BookmarkTree> {
  const [firstTreeInfo, rootFolders] = yield all([
    call(getBookmarkTree, String(options.defExpand)),
    call(getBookmarkChildren, CST.ROOT_ID)
  ])
  return {
    ...firstTreeInfo,
    children: [
      ...R.reject((bookmarkInfo) => {
        const idNumber = Number(bookmarkInfo.id)
        return idNumber === options.defExpand || options.hideRootFolder.includes(idNumber)
      }, rootFolders),
      ...firstTreeInfo.children
    ]
  }
}

type SearchQuery = {| query: string |}
export function* searchBookmarks(searchQuery: SearchQuery): Saga<Array<BookmarkInfo>> {
  const searchResultNodes = yield call(searchBookmarkNodes, searchQuery)
  return R.map(toBookmarkInfo, searchResultNodes)
}

export function* tryGetBookmarkTree(id: string): Saga<?BookmarkTree> {
  try {
    return yield call(getBookmarkTree, id)
  } catch (err) {
    return null
  }
}
