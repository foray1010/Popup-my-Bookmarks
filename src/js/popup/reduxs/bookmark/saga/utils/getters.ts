import * as R from 'ramda'
import {SagaIterator} from 'redux-saga'
import {all, call} from 'redux-saga/effects'

import {Options} from '../../../../../common/types/options'
import {
  getBookmarkChildNodes,
  getBookmarkNodes,
  getI18n,
  searchBookmarkNodes
} from '../../../../../common/utils'
import * as CST from '../../../../constants'
import {BookmarkInfo, BookmarkTree} from '../../../../types'
import {simulateBookmark, toBookmarkInfo} from '../../utils/converters'

export function* getBookmarkInfo(id: string): SagaIterator {
  if (id.startsWith(CST.NO_BOOKMARK_ID_PREFIX)) {
    return simulateBookmark({
      id,
      parentId: id.replace(CST.NO_BOOKMARK_ID_PREFIX, ''),
      title: yield call(getI18n, 'noBkmark'),
      type: CST.BOOKMARK_TYPES.NO_BOOKMARK
    })
  }

  const bookmarkNodes = yield call(getBookmarkNodes, id)
  return toBookmarkInfo(bookmarkNodes[0])
}

export function* getBookmarkChildren(id: string): SagaIterator {
  const bookmarkChildNodes = yield call(getBookmarkChildNodes, id)
  return R.map(toBookmarkInfo, bookmarkChildNodes)
}

export function* getBookmarkTree(id: string): SagaIterator {
  const [bookmarkInfo, bookmarkChildren] = yield all([
    call(getBookmarkInfo, id),
    call(getBookmarkChildren, id)
  ])
  return {
    children: bookmarkChildren.length ?
      bookmarkChildren :
      [yield call(getBookmarkInfo, CST.NO_BOOKMARK_ID_PREFIX + id)],
    parent: bookmarkInfo
  }
}

export function* getBookmarkTrees(
  restTreeIds: Array<string>,
  options: Partial<Options>
): SagaIterator {
  const [firstTree, ...restTrees]: Array<BookmarkTree> = yield all([
    call(getFirstBookmarkTree, options),
    ...restTreeIds.map((id) => call(tryGetBookmarkTree, id))
  ])

  let acc = [firstTree]
  for (const tree of restTrees) {
    // if `tree` is deleted or not belong to this parent anymore,
    // ignore all its children
    const isBreak =
      tree === null ||
      // in case it is root folder that open from root
      (!tree.parent.isRoot && acc[acc.length - 1].parent.id !== tree.parent.parentId)
    if (isBreak) break

    acc = [...acc, tree]
  }
  return acc
}

export function* getFirstBookmarkTree(options: Partial<Options>): SagaIterator {
  const [firstTreeInfo, rootFolders]: [BookmarkTree, Array<BookmarkInfo>] = yield all([
    call(getBookmarkTree, String(options.defExpand)),
    call(getBookmarkChildren, CST.ROOT_ID)
  ])
  return {
    ...firstTreeInfo,
    children: [
      ...R.reject((bookmarkInfo) => {
        const idNumber = Number(bookmarkInfo.id)
        return idNumber === options.defExpand || (options.hideRootFolder || []).includes(idNumber)
      }, rootFolders),
      ...firstTreeInfo.children
    ]
  }
}

interface SearchQuery {
  query: string
}
export function* searchBookmarks(searchQuery: SearchQuery): SagaIterator {
  const searchResultNodes = yield call(searchBookmarkNodes, searchQuery)
  return R.map(toBookmarkInfo, searchResultNodes)
}

export function* tryGetBookmarkTree(id: string): SagaIterator {
  try {
    return yield call(getBookmarkTree, id)
  } catch (err) {
    return null
  }
}
