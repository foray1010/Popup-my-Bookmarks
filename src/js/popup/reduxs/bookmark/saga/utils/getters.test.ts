/* eslint max-lines: 'off' */

import {cloneableGenerator} from '@redux-saga/testing-utils'
import Chance from 'chance'
import * as R from 'ramda'
import {all, call} from 'redux-saga/effects'

import optionsFixture from '../../../../../common/__fixtures__/options.json'
import {
  getBookmarkChildNodes,
  getBookmarkNodes,
  searchBookmarkNodes
} from '../../../../../common/utils'
import * as CST from '../../../../constants'
import {BookmarkTree} from '../../../../types'
import bookmarkNodes from '../__fixtures__/bookmarkNodes'
import bookmarkTrees from '../__fixtures__/bookmarkTrees'
import * as getters from './getters'

const chance = Chance('getters')

describe('getBookmarkInfo', () => {
  test('should get the first bookmark node by id and convert it to bookmarkInfo', () => {
    const generator = getters.getBookmarkInfo(bookmarkNodes[0].id)

    expect(generator.next().value).toEqual(call(getBookmarkNodes, bookmarkNodes[0].id))

    expect(generator.next(bookmarkNodes).value).toMatchSnapshot()

    expect(generator.next().done).toBe(true)
  })
})

describe('getBookmarkChildren', () => {
  test('should get all its children by id and convert all children to bookmarkInfo', () => {
    const generator = getters.getBookmarkChildren(bookmarkNodes[0].id)

    expect(generator.next().value).toEqual(call(getBookmarkChildNodes, bookmarkNodes[0].id))

    expect(generator.next([bookmarkNodes[1], bookmarkNodes[2]]).value).toMatchSnapshot()

    expect(generator.next().done).toBe(true)
  })
})

describe('getBookmarkTree', () => {
  test('get parent and children data at once', () => {
    const generator = getters.getBookmarkTree(bookmarkNodes[0].id)

    expect(generator.next().value).toEqual(
      all([
        call(getters.getBookmarkInfo, bookmarkNodes[0].id),
        call(getters.getBookmarkChildren, bookmarkNodes[0].id)
      ])
    )

    expect(generator.next([bookmarkTrees[0].parent, bookmarkTrees[0].children]).value).toEqual(
      bookmarkTrees[0]
    )

    expect(generator.next().done).toBe(true)
  })
})

describe('getBookmarkTrees', () => {
  // limit to 4 trees, and set parentId to previous tree id
  const correlatedBookmarkTrees = R.take(4, bookmarkTrees).reduce(
    (acc: Array<BookmarkTree>, bookmarkTree) => {
      if (!acc.length) return [bookmarkTree]

      const prevBookmarkTree = R.last(acc)
      if (!prevBookmarkTree) throw new Error('prevBookmarkTree must exist')
      const updatedBookmarkTree = R.set(
        R.lensPath(['parent', 'parentId']),
        prevBookmarkTree.parent.id,
        bookmarkTree
      )
      return [...acc, updatedBookmarkTree]
    },
    []
  )
  const getRestTreeIds = (trees: Array<BookmarkTree>) => R.tail(trees).map((tree) => tree.parent.id)

  const options = optionsFixture
  const restTreeIds = getRestTreeIds(correlatedBookmarkTrees)

  const generator = cloneableGenerator(getters.getBookmarkTrees)(restTreeIds, options)

  expect(generator.next().value).toEqual(
    all([
      call(getters.getFirstBookmarkTree, options),
      ...R.map((id) => call(getters.tryGetBookmarkTree, id), restTreeIds)
    ])
  )

  test('should get all trees', () => {
    const clonedGenerator = generator.clone()

    expect(clonedGenerator.next(correlatedBookmarkTrees).value).toEqual(correlatedBookmarkTrees)

    expect(clonedGenerator.next().done).toBe(true)
  })
  test('should remove tree and its child trees if cannot get that tree', () => {
    for (let i = 1; i < correlatedBookmarkTrees.length; i += 1) {
      const clonedGenerator = generator.clone()

      const partiallyUnrelatedBookmarkTrees = R.set(R.lensIndex(i), null, correlatedBookmarkTrees)
      expect(clonedGenerator.next(partiallyUnrelatedBookmarkTrees).value).toEqual(
        R.take(i, partiallyUnrelatedBookmarkTrees)
      )

      expect(clonedGenerator.next().done).toBe(true)
    }
  })
})
describe('getFirstBookmarkTree', () => {
  test('should get default expanded tree and append root folders to its children except itself', () => {
    const hiddenFolderId = '99999'
    const options = {
      ...optionsFixture,
      defExpand: Number(bookmarkTrees[0].parent.id),
      hideRootFolder: [Number(hiddenFolderId)]
    }

    const generator = getters.getFirstBookmarkTree(options)

    expect(generator.next().value).toEqual(
      all([
        call(getters.getBookmarkTree, String(options.defExpand)),
        call(getters.getBookmarkChildren, CST.ROOT_ID)
      ])
    )

    const rootFolders = [
      {
        ...bookmarkTrees[0].children[0],
        type: CST.BOOKMARK_TYPES.BOOKMARK,
        id: String(chance.integer({min: 2})),
        isRoot: true
      },
      {
        ...bookmarkTrees[0].children[0],
        type: CST.BOOKMARK_TYPES.FOLDER,
        id: String(chance.integer({min: 2})),
        isRoot: true
      },
      {
        ...bookmarkTrees[0].children[0],
        type: CST.BOOKMARK_TYPES.FOLDER,
        id: bookmarkTrees[0].parent.id,
        isRoot: true
      },
      {
        ...bookmarkTrees[0].children[0],
        type: CST.BOOKMARK_TYPES.FOLDER,
        id: hiddenFolderId,
        isRoot: true
      }
    ]
    expect(generator.next([bookmarkTrees[0], rootFolders]).value).toMatchSnapshot()
    expect(generator.next().done).toBe(true)
  })
})

describe('searchBookmarks', () => {
  test('search bookmarks and convert to bookmarkInfo', () => {
    const searchQuery = {query: chance.word()}
    const generator = getters.searchBookmarks(searchQuery)

    expect(generator.next().value).toEqual(call(searchBookmarkNodes, searchQuery))

    expect(generator.next([bookmarkNodes[3], bookmarkNodes[4]]).value).toMatchSnapshot()

    expect(generator.next().done).toBe(true)
  })
})

describe('tryGetBookmarkTree', () => {
  test('run getBookmarkTree', () => {
    const generator = getters.tryGetBookmarkTree(bookmarkNodes[0].id)

    expect(generator.next().value).toEqual(call(getters.getBookmarkTree, bookmarkNodes[0].id))

    expect(generator.next([bookmarkTrees[5]]).value).toEqual([bookmarkTrees[5]])

    expect(generator.next().done).toBe(true)
  })
  test('return null if failed to get bookmark tree', () => {
    const generator = getters.tryGetBookmarkTree(bookmarkNodes[0].id)

    expect(generator.next().value).toEqual(call(getters.getBookmarkTree, bookmarkNodes[0].id))

    // generator.throw should always exists in real world
    if (generator.throw) {
      expect(generator.throw(new Error()).value).toBe(null)
    }

    expect(generator.next().done).toBe(true)
  })
})
