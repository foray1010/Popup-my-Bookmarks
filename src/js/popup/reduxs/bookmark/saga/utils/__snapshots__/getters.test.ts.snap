// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBookmarkChildren should get all its children by id and convert all children to bookmarkInfo 1`] = `
Array [
  Object {
    "iconUrl": "chrome://favicon/http://dobapudol.bd/umowimup",
    "id": "5683",
    "isRoot": false,
    "isSimulated": false,
    "isUnmodifiable": false,
    "parentId": "5863",
    "storageIndex": 4116,
    "title": "ujapernat",
    "type": 0,
    "url": "http://dobapudol.bd/umowimup",
  },
  Object {
    "iconUrl": "chrome://favicon/http://rucophe.sr/pakucug",
    "id": "3980",
    "isRoot": false,
    "isSimulated": false,
    "isUnmodifiable": false,
    "parentId": "",
    "storageIndex": 7243,
    "title": "olevowobe",
    "type": 0,
    "url": "http://rucophe.sr/pakucug",
  },
]
`;

exports[`getBookmarkInfo should get the first bookmark node by id and convert it to bookmarkInfo 1`] = `
Object {
  "iconUrl": "chrome://favicon/http://hipul.va/bihdaeha",
  "id": "8825",
  "isRoot": false,
  "isSimulated": false,
  "isUnmodifiable": false,
  "parentId": "",
  "storageIndex": 7496,
  "title": "woh",
  "type": 0,
  "url": "http://hipul.va/bihdaeha",
}
`;

exports[`getFirstBookmarkTree should get default expanded tree and append root folders to its children except itself 1`] = `
Object {
  "children": Array [
    Object {
      "iconUrl": "http://sidzehod.za/nowocu",
      "id": "5967205401362433",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "4146",
      "storageIndex": 3032,
      "title": "nagcumu",
      "type": 0,
      "url": "http://jefin.tk/se",
    },
    Object {
      "iconUrl": "http://sidzehod.za/nowocu",
      "id": "4488745793355777",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "4146",
      "storageIndex": 3032,
      "title": "nagcumu",
      "type": 2,
      "url": "http://jefin.tk/se",
    },
    Object {
      "iconUrl": "http://sidzehod.za/nowocu",
      "id": "2314",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "4146",
      "storageIndex": 3032,
      "title": "nagcumu",
      "type": 2,
      "url": "http://jefin.tk/se",
    },
    Object {
      "iconUrl": "http://pewojli.lv/pohof",
      "id": "4433",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "2269",
      "storageIndex": 8979,
      "title": "lun",
      "type": 4,
      "url": "http://fopki.at/ugokikto",
    },
    Object {
      "iconUrl": "http://dawtoz.ug/cicop",
      "id": "6984",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "4161",
      "storageIndex": 1272,
      "title": "botverek",
      "type": 2,
      "url": "http://on.ec/tepho",
    },
    Object {
      "iconUrl": "http://sinzov.mx/retu",
      "id": "9824",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "9928",
      "storageIndex": 5093,
      "title": "hi",
      "type": 4,
      "url": "http://jepava.lu/doju",
    },
    Object {
      "iconUrl": "http://remahfur.do/wek",
      "id": "8165",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "5893",
      "storageIndex": 5021,
      "title": "ilo",
      "type": 0,
      "url": "http://jabo.vu/mabheku",
    },
    Object {
      "iconUrl": "http://zahcasmi.sr/hi",
      "id": "9894",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "1047",
      "storageIndex": 4193,
      "title": "hu",
      "type": 4,
      "url": "http://otvuk.nl/wezsi",
    },
    Object {
      "iconUrl": "http://tenula.bh/letnegba",
      "id": "6285",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "1085",
      "storageIndex": 2456,
      "title": "gojku",
      "type": 4,
      "url": "http://sudetu.ac/ajier",
    },
    Object {
      "iconUrl": "http://pedet.se/vibjuwdik",
      "id": "1217",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "2530",
      "storageIndex": 620,
      "title": "mombukfes",
      "type": 0,
      "url": "http://keil.gi/wokuj",
    },
    Object {
      "iconUrl": "http://buk.ax/kusa",
      "id": "9557",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "2384",
      "storageIndex": 8077,
      "title": "ji",
      "type": 0,
      "url": "http://gepfim.md/udojigu",
    },
    Object {
      "iconUrl": "http://va.mx/aso",
      "id": "9551",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "5517",
      "storageIndex": 2338,
      "title": "gu",
      "type": 0,
      "url": "http://iza.gr/ge",
    },
    Object {
      "iconUrl": "http://nula.at/haharud",
      "id": "2421",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "1814",
      "storageIndex": 7501,
      "title": "zuzeboj",
      "type": 2,
      "url": "http://huveimi.lu/dudhe",
    },
    Object {
      "iconUrl": "http://ma.ky/vawujij",
      "id": "3284",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "5594",
      "storageIndex": 7839,
      "title": "ahobaom",
      "type": 0,
      "url": "http://bigeha.pa/bejozkal",
    },
    Object {
      "iconUrl": "http://palaw.gov/din",
      "id": "4335",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "5176",
      "storageIndex": 1909,
      "title": "se",
      "type": 4,
      "url": "http://dab.qa/ha",
    },
    Object {
      "iconUrl": "http://sadcup.ad/ipon",
      "id": "5098",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "8980",
      "storageIndex": 1612,
      "title": "cunnefco",
      "type": 2,
      "url": "http://vogir.bj/diziwef",
    },
    Object {
      "iconUrl": "http://dijolagu.bz/tu",
      "id": "3886",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "9731",
      "storageIndex": 9453,
      "title": "bonawdi",
      "type": 4,
      "url": "http://tuhu.pn/tiwaru",
    },
    Object {
      "iconUrl": "http://uworodim.sz/dud",
      "id": "2298",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "4531",
      "storageIndex": 8972,
      "title": "mipfo",
      "type": 0,
      "url": "http://ra.mx/jozureb",
    },
    Object {
      "iconUrl": "http://fe.nu/hih",
      "id": "5442",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "9273",
      "storageIndex": 347,
      "title": "uduh",
      "type": 2,
      "url": "http://ca.ck/tuna",
    },
    Object {
      "iconUrl": "http://sumjueja.nz/ewnobti",
      "id": "9548",
      "isRoot": true,
      "isSimulated": false,
      "isUnmodifiable": true,
      "parentId": "3059",
      "storageIndex": 8012,
      "title": "juetpoh",
      "type": 0,
      "url": "http://keil.ye/oku",
    },
    Object {
      "iconUrl": "http://haoshe.ga/tohibpow",
      "id": "2595",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "6300",
      "storageIndex": 8468,
      "title": "hamrumba",
      "type": 4,
      "url": "http://zehevduf.fo/cepob",
    },
    Object {
      "iconUrl": "http://jop.by/etenazis",
      "id": "1593",
      "isRoot": false,
      "isSimulated": false,
      "isUnmodifiable": false,
      "parentId": "7659",
      "storageIndex": 8120,
      "title": "cagmakba",
      "type": 2,
      "url": "http://vu.gq/bahafom",
    },
  ],
  "parent": Object {
    "iconUrl": "http://bingola.ke/vef",
    "id": "2724",
    "isRoot": true,
    "isSimulated": false,
    "isUnmodifiable": false,
    "parentId": "2501",
    "storageIndex": 4257,
    "title": "ajuvoj",
    "type": 2,
    "url": "http://niopuloc.ug/fuvugtoh",
  },
}
`;

exports[`searchBookmarks search bookmarks and convert to bookmarkInfo 1`] = `
Array [
  Object {
    "iconUrl": "test-file-stub",
    "id": "9726",
    "isRoot": false,
    "isSimulated": false,
    "isUnmodifiable": true,
    "parentId": "",
    "storageIndex": 6667,
    "title": "nemed",
    "type": 2,
    "url": "",
  },
  Object {
    "iconUrl": "chrome://favicon/http://selke.jo/co",
    "id": "9645",
    "isRoot": false,
    "isSimulated": false,
    "isUnmodifiable": true,
    "parentId": "",
    "storageIndex": -1,
    "title": "mevobo",
    "type": 0,
    "url": "http://selke.jo/co",
  },
]
`;
